crates/typst/src/model/element.rs

```diff
@@ -1,25 +1,105 @@
+use ::std::hash::Hasher;
+use ::typst::model::{Guard, Label, Location};
+use ::typst::syntax::Span;
 use ecow::EcoString;
-use std::any::TypeId;
+use std::any::{Any, TypeId};
 use std::cmp::Ordering;
-use std::fmt::{self, Debug, Formatter};
+use std::fmt::{self, Debug};
+use std::sync::Arc;
 
 use once_cell::sync::Lazy;
 
 use super::{Content, Selector, Styles};
-use crate::diag::SourceResult;
+use crate::diag::{SourceResult, StrResult};
 use crate::eval::{cast, Args, Dict, Func, ParamInfo, Repr, Scope, Value, Vm};
 use crate::util::Static;
 
+pub trait Element: Any + Send + Sync + Debug + Repr + 'static {
+    /// Get the element data for this element.
+    fn data(&self) -> ElementData;
+
+    /// Get the element's span.
+    ///
+    /// May be detached if it has not been set.
+    fn span(&self) -> Span;
+
+    /// Sets the span of this element.
+    fn set_span(&mut self, span: Span);
+
+    /// Get the element's location.
+    fn location(&self) -> Option<Location>;
+
+    /// Sets the location of this element.
+    fn set_location(&mut self, location: Location);
+
+    /// Get the element's label.
+    fn label(&self) -> Option<Label>;
+
+    /// Sets the label of this element.
+    fn set_label(&mut self, label: Label);
+
+    /// Checks whether the element is guarded by the given guard.
+    fn is_guarded(&self, guard: Guard) -> bool;
+
+    /// Pushes a guard onto the element.
+    fn push_guard(&mut self, guard: Guard);
+
+    /// Whether the element is pristine.
+    fn is_pristine(&self) -> bool;
+
+    /// Mark the element as having been prepared.
+    fn mark_prepared(&mut self);
+
+    /// Whether this element needs preparations.
+    fn needs_preparation(&self) -> bool;
+
+    /// Whether this element has been prepared.
+    fn is_prepared(&self) -> bool;
+
+    /// Dynamically hash the element.
+    fn dyn_hash(&self, hasher: &mut dyn Hasher);
+
+    /// Dynamically compare the element.
+    fn dyn_eq(&self, other: &Content) -> bool;
+
+    /// Get the field with the given field ID.
+    fn field(&self, id: u8) -> Option<Value>;
+
+    /// Set the fields of the element.
+    fn set_field(&mut self, id: u8, value: Value) -> StrResult<()>;
+
+    /// Dynamically clone the element.
+    fn dyn_clone(&self) -> Arc<dyn Element>;
+
+    /// Get the fields of the element.
+    fn fields(&self) -> Dict;
+}
+
 /// A document element.
 #[derive(Copy, Clone, Eq, PartialEq, Hash)]
-pub struct Element(Static<NativeElementData>);
+pub struct ElementData(Static<NativeElementData>);
```

Comment:
If we merge the `Element` and `NativeElement` traits, this could be renamed back. I liked the old name much more. I had always trouble with naming these things and it took a while to land at the current naming scheme, with which I am very happy (Func/Element/Type, NativeFunc/NativeElement/NativeType, NativeFuncData/NativeElementData/NativeTypeData).

Reply:
It still cannot be due to the `Set` trait not being object safe, I had tried again to merge them but it's just not possible :(

Reply:
I lied 💀